pkg/controller/deorbit/controller.go

```diff
@@ -0,0 +1,149 @@
+
+	"github.com/sapcc/kubernikus/pkg/api/models"
+	"github.com/sapcc/kubernikus/pkg/apis/kubernikus/v1"
+	"github.com/sapcc/kubernikus/pkg/controller/base"
+	"github.com/sapcc/kubernikus/pkg/controller/config"
+	"github.com/sapcc/kubernikus/pkg/controller/metrics"
+	"github.com/sapcc/kubernikus/pkg/util"
+)
+
+const (
+	DeorbiterFinalizer = "deorbiter"
+
+	// The longest time a single worker will be blocked, while waiting for the clean
+	// up of OpenStack resources. After this timeout, deobiting will be queued again.
+	// It's a safeguard against having all workers hanging indefinetly and congesting
+	// the queue.
+	UnblockWorkerTimeout = 10 * time.Minute
+
+	// If the customer's apiserver is unreachable for this duration, we assume it is
+	// already decomissioned, permanently damaged or was never created successfully
+	// in the first place.
+	APIUnavailableTimeout = 2 * time.Minute
+
+	// After this duration, the deorbiter will utimately give up retrying. This will
+	// potentially leave debris in the customer's project in the form of volumes,
+	// load balancers or floating IPs.
+	DeorbitHangingTimeout = 24 * time.Hour
+)
+
+type DeorbitReconciler struct {
+	config.Clients
+
+	Recorder record.EventRecorder
+	Logger   log.Logger
+}
+
+func NewController(factories config.Factories, clients config.Clients, recorder record.EventRecorder, logger log.Logger) base.Controller {
+	logger = log.With(logger, "controller", "deorbit")
+
+	var reconciler base.Reconciler
+	reconciler = &DeorbitReconciler{clients, recorder, logger}
+	reconciler = &base.LoggingReconciler{reconciler, logger}
+	reconciler = &base.InstrumentingReconciler{
+		reconciler,
+		metrics.DeorbitOperationsLatency,
+		metrics.DeorbitOperationsTotal,
+		metrics.DeorbitSuccessfulOperationsTotal,
+		metrics.DeorbitFailedOperationsTotal,
+	}
+	return base.NewController(factories, clients, reconciler, logger)
+}
+
+func (d *DeorbitReconciler) Reconcile(kluster *v1.Kluster) (bool, error) {
+	switch kluster.Status.Phase {
+	case models.KlusterPhaseRunning:
+		return false, util.EnsureFinalizerCreated(d.Kubernikus.KubernikusV1(), kluster, DeorbiterFinalizer)
+	case models.KlusterPhaseTerminating:
+		if kluster.HasFinalizer(DeorbiterFinalizer) {
+			if err := d.deorbit(kluster); err != nil {
+				return false, err
+			}
+			return false, util.EnsureFinalizerRemoved(d.Kubernikus.KubernikusV1(), kluster, DeorbiterFinalizer)
+		}
+	}
+	return false, nil
+}
+
+func (d *DeorbitReconciler) deorbit(kluster *v1.Kluster) (err error) {
+	// The following channel is used to abort the deorbiting after a certain
+	// time. It is required because the deorbit wait-functions block indefinetly or
+	// until the stop channel is closed. This is used to unblock the workqueue.
+	done := make(chan struct{})
+	defer close(done)
+
+	timer := time.NewTimer(UnblockWorkerTimeout)
+	defer timer.Stop()
+
+	go func() {
+		<-timer.C
+		d.Logger.Log("msg", "timeout waiting. unblocking the worker routine")
+		close(done)
```

Comment:
Together with line 81 we have a guaranteed panic here when the timeout is hit I think  (close of already closed channel).

Reply:
Hm. You are right. I fixed this now but it feels like there's a simpler way to do that.